File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.scala
import scala.collection.mutable._

object Solution {
    def lengthOfLIS(nums: Array[Int]): Int = {
        val list: ListBuffer[Int] = ListBuffer()

        nums.foreach { num => 
            val i = binSearch(list, num)

            if (i == list.length) list += num
            else list(i) = num
        }

        list.length
    }

    def binSearch(list: ListBuffer[Int], target: Int): Int = {
        var s = 0
        var e = list.length - 1

        while (s <= e) {
            val m = (s + e) / 2
            val v = list(m)

            if (v >= target) {
                e = m - 1
            } else {
                s = m + 1
            }
        }

        s
    }
}

File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        List<Integer> list = new ArrayList<>();
        int res = 0;
        
        for(int n : nums) {
            int i = lowerBound(list, n);
            
            if (i == list.size()) list.add(n);
            else list.set(i, n);
            
            res = Math.max(res, list.size());
        }
        
        return res;
    }
    
    int lowerBound(List<Integer> list, int t) {
        int l = 0;
        int r = list.size() - 1;
        
        while(l <= r) {
            int m = l + (r - l) / 2;
            if (list.get(m) >= t) r = m - 1; 
            else l = m + 1;
        }
        
        return l;
    }
}